### ISSUE TITLE
Can we support wifi,mobile network condition

### ISSUE THREAD

**Issue by mahuafactory** [yigit/android-priority-jobqueue#62]
some job must work when the wifi connection

**Comment by mahuafactory** [comment]
Can I join you and push some code?

**Comment by kingargyle** [comment]
@mahuafactory just for the repository, and submit a pull request for any changes you make. @yigit will review it.  it is the best way to make sure your changes get into the main branch.

**Comment by yigit** [comment]
Do you have a fork for this already?
I would like to see how you are implementing it. 

With the JobScheduler integration, I think we'll need more of these. (phone charging etc, whatever power manager provides). We may even consider supporting custom logic although it makes it even more complex and I'm not sure about the added value.

With the current setup, at least how the NonPersistent Queue persists jobs should change to allow more queries.

**Comment by bangarharshit** [comment]
+1 for the awesome suggestion. I think automatic sync on wifi is cool idea especially for handling big image and video uploads/downloads while certainly they shouldn't be done on 2g/3g/ or any mobile pack.

@yigit I am cancelling the job in retry callback when I  am getting networkexception. Do you think it can be part of library.

Also, android's connectivityreceiver is not that great. It doesn't actually tell if you are actually connected or not. Do you guys suggest some workaround?

**Comment by yigit** [comment]
Hi,
Yes this feature will arrive also with gcm network manager support.

Current JobManager codebase / API hit some complexity limit where implementing new features became hard and requiring API breaking (e.g. this change requires changing JobQueue API).
So I decided to rewrite internals with a new architecture to make it more scalable in the future.
see here:
https://github.com/yigit/android-priority-jobqueue/wiki/JobManager-2.0---Roadmap

This branch has the re-write.
https://github.com/yigit/android-priority-jobqueue/tree/message-queue

I'm almost done with the initial step (the hard part) and almost all tests passing. The only major API I changed is the `addJob` where it won't return a `long` id anymore, instead, Job's will have `UUID` that is reliably unique across all JobQueues. Once all tests pass, I'll change the JobQueue API to allow more constraints (maybe even custom constraints but no promises on that).

**Comment by bangarharshit** [comment]
Awesome. Yeah GCM network manager support will be awesome. I wrote a small version using tape but it was not at all flexible and we certainly ignored cases like Doze and so we are switching to your library. Your design decisions are interesting and would love to discuss. UUID is cool and we use it everywhere. Do you think instead of generating UUIDs internally you can give  this responsibility to clients who are using the library. Pro - It will be useful to maintain idempotence of the job. If the client is submitting job on button click and accidentally clicks button twice (it happens to me frequently if I remain on same screen) he will submit two jobs. If the UUID is external and assuming he is doing it correctly he will just generate one uuid - And then inside priority queue you can check if there is a job, just update or discard. Con - Most of the clients will probably implement it correctly. Otherwise job object's hashcode will be fine.

**Comment by yigit** [comment]
I thought about letting client define an ID generator but I want them to be generated when Job is created (not necessarily added). 
Adding two jobs with the same ID is a tricky case. The first job might have started by the time second add arrived, what is the JobManager supposed to do? I want to solve it via an API, there is even a PR for it: https://github.com/yigit/android-priority-jobqueue/pull/82
The API is not final yet but I'm hoping to provide something where you can specify what the behavior should be if the previous job is running etc.
Today, I finally made all tests pass in the message queue branch, after doing some cleanup and documentation, I'll release a snapshot.

**Comment by bangarharshit** [comment]
Great. Thanks. :100:

**Comment by yigit** [comment]
Another thing here is that I think the API should provide a wall-clock timeout.
There are simply phones out there which will never see a wireless network.

Right now JobManager does not use wall clock (so after restart, everything is 0) but I'm planning to change it for 2.0.

tl;dr; This API should allow the following case:

`Run this job on WIFI, if it is not run after 48 hours, run it on any network.`

**Comment by mahuafactory** [comment]
In China,we can not use google service.There is a Great Fire Wall that build by government to prevent us access the site like google,facebook,tiwtter.Sadly!!!

**Comment by yigit** [comment]
Hi, I've uploaded snapshot artifacts from v2 branch with this feature. 

You can access it via adding snapshot repository into build gradle:

    dependencies {
        maven {
            url 'https://oss.sonatype.org/content/repositories/snapshots/'
        }
    }

and changing your version to `2.0.0-SNAPSHOT`. Lmk if you can give it a try.
The `Params` object has the following network related methods:

    requireWifiNetworkWithTimeout, requireWifiNetwork, requireWifiWithTimeout, requireNetwork

This allows you create a job that requires a WIFI for some timeout and after that accept any network.

    new Param(1).requireWifiWithTimeout(60 * 60 * 1000 /*1 hour*/) .requireNetwork()

**Comment by yigit** [comment]
I don't have the integration with JobScheduler  (or alarm manager) yet but that'll come before 2.0 is released.

**Comment by yigit** [comment]
available on v2.
